### PR TITLE
Protect

### PR DESCRIPTION
**Pull request by jasonpartyka**
Per our discussion at DrupalCon, this is a refactoring of #123 to be generalized and in a submodule.

@agentrickard

### CODE REVIEW

**Comment by jasonpartyka** [comment]
I can't figure out how to make the tests for the submodule pass in Travis. 

I can run them locally like so:
`vendor/bin/phpunit -c core modules/contrib/workbench_access/workbench_access_protect/tests/Functional/TaxonomySchemeUITest.php`

File: .travis.yml [hunk 1]
```diff
@@ -77,3 +77,4 @@ script:
   - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Functional" --php `which php` --url http://localhost:8080 "workbench_access"
   - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Kernel" --php `which php` --url http://localhost:8080 "workbench_access"
   - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Unit" --php `which php` --url http://localhost:8080 "workbench_access"
+  - vendor/bin/phpunit -c core "modules/contrib/workbench_access/workbench_access_protect/tests/Functional/TaxonomySchemeUITest.php"
```
Comment (agentrickard) [inline .travis.yml]: I would use the same method as above, which means annotating the tests, I believe.

File: workbench_access_protect/src/Access/DeleteAccessCheckInterface.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Drupal\workbench_access_protect\Access;
+
+use Drupal\Core\Entity\EntityInterface;
+
+/**
+ * Class TaxonomyDeleteAccessCheck.
+ *
+ * @package Drupal\workbench_access\Access
```
Comment (agentrickard) [inline workbench_access_protect/src/Access/DeleteAccessCheckInterface.php]: remove

File: workbench_access_protect/tests/Functional/TaxonomySchemeUITest.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+use Drupal\Tests\BrowserTestBase;
+use Drupal\taxonomy\Entity\Vocabulary;
+use Drupal\Tests\workbench_access\Traits\WorkbenchAccessTestTrait;
+use Drupal\taxonomy\Entity\Term;
+use Drupal\workbench_access\WorkbenchAccessManagerInterface;
+
+/**
+ * Defines a class for testing the UI to create and configure schemes.
+ *
+ * @group workbench_access
```
Comment (agentrickard) [inline workbench_access_protect/tests/Functional/TaxonomySchemeUITest.php]: `@group workbench_access_protect`

File: workbench_access_protect/workbench_access_protect.info.yml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 name: Workbench Access Protect
 type: module
-description: Prevents deletion
+description: Enable `Workbench Access Protect` to prevent entities from being
```
Comment (agentrickard) [inline workbench_access_protect/workbench_access_protect.info.yml]: Incomplete sentence. Does not need module name.

File: workbench_access_protect/workbench_access_protect.module [hunk 5]
```diff
@@ -6,22 +6,37 @@
...
 
 /**
  * Implements hook_entity_access().
  */
 function workbench_access_protect_entity_access(EntityInterface $entity, $op, AccountInterface $account) {
   // Return net result of all enabled access schemes. If one scheme allows
   // access, then it is granted.
+
+  /** @var Drupal\workbench_access_protect\Access\DeleteAccessCheck $protect */
+  $protect = \Drupal::service('workbench_access_protect.delete');
```
Comment (agentrickard) [inline workbench_access_protect/workbench_access_protect.module]: We only need to load the service if $op == delete.

File: .travis.yml (hunk 1 above)
Comment (jasonpartyka) [reply .travis.yml]: Changed to the same method, still failing.

**Comment by jasonpartyka** [comment]
So, there's another thing. The "taxonomy" and "taxonomy_term" are actually two different entities (I may have those entity-type-names slightly wrong, I don't have a code base up as I'm writing this). 

I noticed that my code, as written, still allows you to delete a vocabulary if terms are being actively used. I didn't notice it at first because a user was assigned to the vocabulary as whole -- so that triggered a state where you couldn't delete it. When I removed said user, and no other users were assigned to the vocabulary as whole, I could now delete the vocabulary. This is because of the different entity types.

Are you aware of any data structure that defines this 'parent-child' relationship that we could use?

**Comment by agentrickard** [comment]
Right, 'vocabulary' is the parent and 'taxonomy_term' is the child. If I recall correctly, taxonomy_term is a bundle of vocabulary. The intent of checking the bundles of the item being deleted was to catch this type of inheritance.

On delete check for a Vocabulary, we should be ale to loop through its child bundles and then see if any of them are used for access control. If so, we would deny access to the delete operation.

THis is what the `hasBundles()` and `getBundles()` methods are designed to do.

**Comment by jasonpartyka** [comment]
So, in this method (excuse the debugging code assists):
```
  public function hasBundles(EntityInterface $entity) {
    /** @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface $service */
    $service = \Drupal::service('entity_type.bundle.info');
    $t = $service->getAllBundleInfo();
    $info = $service->getBundleInfo($entity->getEntityTypeId());
    return $info;
  }
```

The `$t` variable has them as two separate entries:

image

Am I misinterpreting this? To me, this looks like 'taxonomy_term' and 'taxonomy_vocabulary' are actually different entity types. Is there some way I can extract this linkage from the Drupal API in a generic way? I've been searching to find a way, but no luck.

**Comment by agentrickard** [comment]
Well, the  information we need might not be loaded onto the bundle itself. We may have to use a lookup similar to that in `Vocabulary::postDelete()` which loads all fields that reference the deleted vocabulary.

```
    // Load all Taxonomy module fields and delete those which use only this
    // vocabulary.
    $field_storages = entity_load_multiple_by_properties('field_storage_config', ['module' => 'taxonomy']);
    foreach ($field_storages as $field_storage) {
...
```

There may or may not be existing methods for "get me all child entities of this parent entity."

**Comment by jasonpartyka** [comment]
I just pushed a commit, but it's not quite ready yet.

What I still need to figure out is how to limit the items being checked to just the ones being used for active access control. Haven't investigated that yet, but @agentrickard do you have a suggestion?

**Comment by agentrickard** [comment]
That code snippet I posted yesterday will find fields, and I think we also have a registry of them in our config. Unfortunately, that config registry is specific to each config type (taxonomy or menu) so it might not be abstract-able.

**Comment by jasonpartyka** [comment]
I really didn't explain much about what my push does. Sorry about that.

So, what I just committed does successfully check for the situation I described above (the taxonomy_vocabulary and taxonomy_term entity-types). However, it's now checking way too aggressively, as it's checking on reference fields that are not even allowed to reference against the controlling entity. It's then getting false positives, this code is denying you the ability to delete if it's in any entity reference field.

So, put another way in terms of taxonomy terms (but actually abstracted in reality), if I can tell that this reference field is not even allowed to reference terms in the controlling vocabulary, then I want to skip altogether and not bother checking.

**Comment by jasonpartyka** [comment]
So, in looking at the entity structure for taxonomies, it looks like we may have all the information there, it's just named in a non-generic manner.

I wanted to see what `$this->entityTypeManager->getStorage('access_scheme')->loadMultiple()` returned (in `DeleteAccessCheck::isAssignedToContent`):

image

image

I'll see if this could work for menu assignments as well, but it's possible that the `vocabularies` index should have just been named `ids`. This could break BC, so maybe a simple fix is just to add an `ids` index?

**Comment by jasonpartyka** [comment]
@agentrickard So, I added an `ids` index to the configuration access in the `workbench_access` module and I was able to do only check if the entity in question was part of a bundle being actively used for access control. It's kind of redundant, however, as it's the same as the `vocabularies` and `menus` index.

**Comment by jasonpartyka** [comment]
Also, not sure at the moment if the `menus` changes are correct. That's mostly because I'm not sure how menus are used for access control. Is it the same like taxonomy, where it's an entity reference to the menu entity?

@agentrickard

**Comment by jasonpartyka** [comment]
@agentrickard  Just following up on this. I had a question in my last comment, and still having issues running the tests since they're in a submodule.

**Comment by agentrickard** [comment]
Menus are indeed an entity reference, though they are a little different than taxonomies. 

What kind of testing issues?

**Comment by jasonpartyka** [comment]
The test runner can't find the tests in the submodule.

**Comment by jasonpartyka** [comment]
It doesn't even look like it can run the `workbench_access` ones correctly. I'm not familiar with travis at all, and just a little bit with the drupal test runner, so i'm at a loss as to how to correct this.

**Comment by agentrickard** [comment]
All of the Workbench Access test fails are because you added the `ids` config element without updating the module schema.

```
    Drupal\Tests\workbench_access\Kernel\ConfigDependenciesTest::testViewsFieldDependencies
    Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for
    workbench_access.access_scheme.editorial_section with the following errors:
    workbench_access.access_scheme.editorial_section:scheme_settings.ids
    missing schema
```

See https://www.drupal.org/docs/8/api/configuration-api/configuration-schemametadata

For the new tests, you have to add a `@group` annotation to the class comment. E.g.

```
/**
 * Tests for the access by role form.
 *
 * @group workbench_access
 */
class AccessByRoleFormTest extends BrowserTestBase {
```

**Comment by jasonpartyka** [comment]
Doh!  Now I see the error.

**Comment by jasonpartyka** [comment]
So the group class comment is there:

https://github.com/jasonpartyka/workbench_access/blob/protect/workbench_access_protect/tests/Functional/TaxonomySchemeUITest.php

```
/**
 * Defines a class for testing the UI to create and configure schemes.
 *
 * @group workbench_access_protect
 */
class TaxonomySchemeUITest extends BrowserTestBase {
```

File: workbench_access_protect/src/Access/DeleteAccessCheck.php [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+use Drupal\field\Entity\FieldStorageConfig;
+use Drupal\Core\Routing\CurrentRouteMatch;
+use Drupal\workbench_access\UserSectionStorage;
+use Drupal\Core\Entity\EntityFieldManager;
+use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\workbench_access\Entity\AccessSchemeInterface;
+
+/**
+ * Class TaxonomyDeleteAccessCheck.
+ */
```
Comment (agentrickard) [inline workbench_access_protect/src/Access/DeleteAccessCheck.php]: Needs a `@group` annotation.

File: .travis.yml [hunk 7]
```diff
@@ -77,3 +77,4 @@ script:
   - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Functional" --php `which php` --url http://localhost:8080 "workbench_access"
   - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Kernel" --php `which php` --url http://localhost:8080 "workbench_access"
   - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Unit" --php `which php` --url http://localhost:8080 "workbench_access"
+  - php core/scripts/run-tests.sh --suppress-deprecations --verbose --color --concurrency 4 --types "PHPUnit-Unit" --php `which php` --url http://localhost:8080 "workbench_access_protect"
```
Comment (agentrickard) [inline .travis.yml]: These should call `PHPUnit-Functional` and not `PHPUnit-Unit`, since you wrote Functional tests.

**Comment by agentrickard** [comment]
What are these IDs for? Do we really need them saved or can we drive them when needed?

**Comment by jasonpartyka** [comment]
That is storing the parent entity id so that we can check on the parent entity as a whole on delete time. It's used in `\Drupal\workbench_access_protect\Access\DeleteAccessCheck::isDeleteAllowedBundle`.

**Comment by jasonpartyka** [comment]
So, I wanted to make sure you were OK with the IDs being added to the schema like that. I still need to get an update in there somewhere for existing installs.

**Comment by agentrickard** [comment]
I think that adding IDs is a bad idea, simply because they are content ids (and therefore not assumed to be identical across installations).

We can store / export machine names/ids (like vocabulary names), but exporting term or menu ids means that we're trying to export content ids, which can change from LOCAL to DEV to PROD.

So I would try to achieve the same thing without exporting numeric ids.

(If I have misunderstood and these are machine ids, then it's fine.)

**Comment by jasonpartyka** [comment]
> I think that adding IDs is a bad idea, simply because they are content ids (and therefore not assumed to be identical across installations).
> 
> We can store / export machine names/ids (like vocabulary names), but exporting term or menu ids means that we're trying to export content ids, which can change from LOCAL to DEV to PROD.
> 
> So I would try to achieve the same thing without exporting numeric ids.
> 
> (If I have misunderstood and these are machine ids, then it's fine.)

Yes, these are the machine IDs of the 'parent' entity. That's what needs to be stored in order to determine if the vocab is in in active use. Maybe a different name is in order so that it's less confusing?

**Comment by agentrickard** [comment]
I think `parent_ids` is probably fine, with documentation that indicates what gets stored there.

**Comment by jasonpartyka** [comment]
I figured out the problem with why the `workbench_access_protect` tests weren't running, but now there's another error. The tests are generating HTML -- do you know if there's a way I can get the HTML from Travis?

**Comment by agentrickard** [comment]
No. But you can run the command locally to see the output:

`php core/scripts/run-tests.sh --suppress-deprecations --verbose --browser --color --types "PHPUnit-Functional"  --url http://localhost:8080 "workbench_access_protect"`

Make sure you change the `--url` to your local dev url.

**Comment by jasonpartyka** [comment]
@agentrickard So, something is not correct with the test setup() methods. I'm trying to get a faster way to run the tests (currently, within docker which is super-slow for functional tests) so I can figure out what's incorrect.

What is your testing setup tools like?

File: workbench_access_protect/src/Access/DeleteAccessCheck.php [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+      $hasAccessControlMembers = $this->doesTermHaveMembers($entity);
+      $assigned_content = $this->isAssignedToContent($entity);
+
+      /*
+       * If this entity does not have users assigned to it for access
+       * control, and the term is not assigned to any pieces of content,
+       * it is OK to delete it.
+       */
+      if ($hasAccessControlMembers || $assigned_content) {
```
Comment (biz123) [inline workbench_access_protect/src/Access/DeleteAccessCheck.php]: @jasonpartyka the problem I was seeing can be solved by changing this conditional so the operator is `&&` and not `||`.

I was seeing all vocabularies set not delete-able and tweaking that makes it so only vocabularies that are used for access control and have content are protected, which seems like what we want.

**Comment by agentrickard** [comment]
Merging to bring this branch up-to-date with the fork. Still needs work.

**Comment by agentrickard** [comment]
@jasonpartyka I also just merged 8.x-1.x into this branch, so you'll want to update your fork.

**Pull request merged**